metadata/moa/scripts/nbconverted/3.align-moa.py

```diff
@@ -0,0 +1,531 @@
+            .dropna()
+            .str.split("|")
+            .tolist(),
+            index=map_df.dropna(subset=[broad_id_col]).index,
+        )
+        .stack()
+        .reset_index()
+    )
+    broad_split_df.columns = ["map_index", "_", "broad_id_unique"]
+
+    map_long_df = (
+        map_subset_df.merge(
+            broad_split_df.loc[:, ["map_index", "broad_id_unique"]],
+            left_index=True,
+            right_on="map_index",
+            how="left",
+        )
+        .reset_index(drop=True)
+        .loc[:, ["InChIKey14", "broad_id_unique"]]
+        .drop_duplicates()
+        .assign(broad_date=broad_id_col)
+        .dropna()
+        .reset_index(drop=True)
+    )
+
+    return map_long_df
+
+
+# ## Step 1: Define InChIKey14-to-Broad ID Map
+
+# In[4]:
+
+
+mapping_col = "InChIKey14"
+
+repurposing_col = "broad_id_20170327"
+broad_id_other_dates = ["20170327", "20180516", "20180907", "20200324"]
+
+
+# In[5]:
+
+
+# Load platemap data
+file = pathlib.PurePath("../platemaps/broad_sample_info.tsv")
+all_platemap_df = pd.read_csv(file, sep="\t")
+
+print(all_platemap_df.shape)
+all_platemap_df.head()
+
+
+# In[6]:
+
+
+# How many unique broad IDs?
+len(all_platemap_df.broad_id.unique())
+
+
+# In[7]:
+
+
+# Load MOA map
+map_file = pathlib.PurePath("clue/broad_id_map.csv")
+map_df = pd.read_csv(map_file)
+
+print(map_df.shape)
+map_df.head()
+
+
+# In[8]:
+
+
+# Create MOA map long mapping every Broad ID to InChIKey14
+all_broad_map = []
+for date in broad_id_other_dates:
+    broad_id_col = f"broad_id_{date}"
+    deprecated_broad_id_col = f"deprecated_broad_id_{date}"
+
+    select_cols = [mapping_col, broad_id_col, deprecated_broad_id_col]
+    map_subset_df = map_df.loc[:, select_cols]
+
+    broad_long_df = split_pipe_broad_id(map_subset_df, broad_id_col)
+    all_broad_map.append(broad_long_df)
+
+    if date != "20170327":
+        deprecated_broad_long_df = split_pipe_broad_id(
+            map_subset_df, deprecated_broad_id_col
+        )
+        all_broad_map.append(deprecated_broad_long_df)
+
+# Combine full broad_id map
+all_broad_map = (
+    pd.concat(all_broad_map)
+    # This step drops InChIKeys and Broad_ID combos that exist in multiple dates
+    .drop_duplicates(subset=[mapping_col, "broad_id_unique"]).reset_index(drop=True)
+)
+
+print(all_broad_map.shape)
+all_broad_map.head()
+
+
+# In[9]:
+
+
+# Are there Duplicate InChIKey14s?
+inchi_counts = all_broad_map.loc[:, mapping_col].squeeze().value_counts()
+inchi_counts.hist()
+
+
+# In[10]:
+
+
+# Show an example of the most egregious multi-mapping InChiKey14
+all_broad_map.loc[all_broad_map.loc[:, mapping_col] == inchi_counts.index[0], :]
+
+
+# ## Step 2: Perform the Merge with the Platemap Data
+# 
+# This step identifies the InChIKey14 map for the Cell Painting Repurposing Hub Data Subset.
+
+# In[11]:
+
+
+merged_platemap_df = (
+    all_platemap_df.merge(
+        all_broad_map, left_on="broad_id", right_on="broad_id_unique", how="left"
```

Comment:
I never realized that the column names of the data frames could be different in the two data frames :)